clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/common/AzureUtilities.groovy

```diff
@@ -112,6 +103,37 @@ class AzureUtilities {
     resourceId.split(PATH_SEPARATOR).last()
   }
 
+  // For id = "/subscriptions/***-***-***/resourceGroups/***/providers/Microsoft.Network/networkInterfaces/nic1/ipConfigurations/ipconfig1"
+  //   this method return "nic1"
+  static String getResourceNameFromId(String id) {
+    if (id == null) {
+      return null
+    }
+
+    def vals = id.split(PATH_SEPARATOR)
+
+    if (vals.length > 8) {
+      return vals[8] // see vals.findIndexOf { it == "Microsoft.Network"} + 2
```

Comment:
In this current implementation, the assumption is that the 8th position of the resource Id will always be the name of the main/top-level resource. However, I'm not sure if that will always be the case. In fact looking at the sample given in the comment above the method, position 8 in the _vals_ array would be "ipConfigurations", not "nic1", which is the resource name we're actually looking for. (arrays are zero-based indexes for arrays)
In the previous implementation of this method, you were looking for the name relative to the location of the Resource Provider name. I think this is a better approach as it is more certain that the main/top-level resource will follow the _<ResourceProvider>/<ResourceType>/<ResourceName>_ pattern in the resource id. The only problem before was that you had hard-coded the resource provider name to a specific resource provider (Microsoft.Network) rather than getting via parameter passed in by the caller. 

Also, we should add tests for this method and the getResourceTypeFromId method below